pkg/importer/imageio-datasource.go

```diff
@@ -18,626 +18,110 @@ package importer
-			_, err = transfersService.ImageTransferService(itID).Finalize().Send()
-			if err != nil {
-				_, cancelErr := transfersService.ImageTransferService(itID).Cancel().Send()
-				if cancelErr != nil {
-					klog.Errorf("Unable to finalize or cancel transfer! Disk may remain locked until inactivity timeout.\n%v\n%v", err, cancelErr)
-				}
-			}
-		}
-	}
-	if is.connection != nil {
-		err = is.connection.Close()
-	}
-	is.cancelLock.Lock()
-	if is.cancel != nil {
-		is.cancel()
-		is.cancel = nil
-	}
-	is.cancelLock.Unlock()
-	return err
+	return &imageIODataSource{
+		download: download,
+		cancel:   cancel,
+	}, nil
 }
 
-func (is *ImageioDataSource) pollProgress(reader *util.CountingReader, idleTime, pollInterval time.Duration) {
-	count := reader.Current
-	lastUpdate := time.Now()
-	for {
-		if count < reader.Current {
-			// Some progress was made, reset now.
-			lastUpdate = time.Now()
-			count = reader.Current
-		}
-
-		if time.Until(lastUpdate.Add(idleTime)).Nanoseconds() < 0 {
-			is.cancelLock.Lock()
-			if is.cancel != nil {
-				// No progress for the idle time, cancel http client.
-				is.cancel() // This will trigger dp.ctx.Done()
-			}
-			is.cancelLock.Unlock()
-		}
-		select {
-		case <-time.After(pollInterval):
-			continue
-		case <-is.ctx.Done():
-			return // Don't leak, once the transfer is cancelled or completed this is called.
-		}
-	}
+type imageIODataSource struct {
+	download ovirtclient.ImageDownload
+	cancel   context.CancelFunc
 }
 
-func createImageioReader(ctx context.Context, ep string, accessKey string, secKey string, certDir string, diskID string) (io.ReadCloser, uint64, *ovirtsdk4.ImageTransfer, ConnectionInterface, error) {
-	conn, err := newOvirtClientFunc(ep, accessKey, secKey)
-	if err != nil {
-		return nil, uint64(0), nil, conn, errors.Wrap(err, "Error creating connection")
-	}
-
-	it, total, err := getTransfer(conn, diskID)
-	if err != nil {
-		return nil, uint64(0), it, conn, err
-	}
-
-	// Use the create client from http source.
-	client, err := createHTTPClient(certDir)
-	if err != nil {
-		cancelTransfer(conn, it)
-		return nil, uint64(0), it, conn, err
-	}
-	transferURL, available := it.TransferUrl()
-	if !available {
-		cancelTransfer(conn, it)
-		return nil, uint64(0), it, conn, errors.New("Error transfer url not available")
-	}
-
-	req, err := http.NewRequest("GET", transferURL, nil)
-	req = req.WithContext(ctx)
-
-	resp, err := client.Do(req)
-	if err != nil {
-		cancelTransfer(conn, it)
-		return nil, uint64(0), it, conn, errors.Wrap(err, "Sending request failed")
-	}
-	if resp.StatusCode != http.StatusOK {
-		cancelTransfer(conn, it)
-		return nil, uint64(0), it, conn, errors.Errorf("bad status: %s", resp.Status)
-	}
-
-	if total == 0 {
-		// The total seems bogus. Let's try the GET Content-Length header
-		total = parseHTTPHeader(resp)
-	}
-	countingReader := &util.CountingReader{
-		Reader:  resp.Body,
-		Current: 0,
+func (i *imageIODataSource) Info() (ProcessingPhase, error) {
```

Comment:
The info functions purpose is to gather information about the source, in particular about the size of the image we are about to download. Once we have that information we can determine if we can even complete the transfer (not enough space would be a problem for instance)

Reply:
In case of oVirt the space is known until you actually start the download. The oVirt engine cannot always provide reliable information on the image size. Therefore, in some cases the Content-Length header of the HTTP response is used to determine the size.